### ISSUE TITLE
[CLARIFICATION REQUEST] Chapter 16 Beam Search

### ISSUE THREAD

**Issue by ethenkaufmann** [ageron/handson-ml2#541]
Hi @ageron,

I was reading chapter 16, everything is beautifully explained. 
Just one thing I did not understand about beam-search is: is it used during training, or will be applied once training is finished during the inference?

In the following code snippet, how do one define the ```start_tokens``` argument, I had a look at the ```tensorflow's``` API and did not understand it?
Would something like ```start_tokens = tf.full([batch_size], 0, tf.int32)```, where **batch_size** is **32** and **SOS** id is *0*, work?

```python 
beam_width = 10
decoder = tfa.seq2seq.beam_search_decoder.BeamSearchDecoder(
    cell=decoder_cell, beam_width=beam_width, output_layer=output_layer)
decoder_initial_state = tfa.seq2seq.beam_search_decoder.tile_batch(
    encoder_state, multiplier=beam_width)
outputs, _, _ = decoder(
    embedding_decoder, start_tokens=start_tokens, end_token=end_token,
    initial_state=decoder_initial_state)
```

I am not sure if this is the correct way, but I modified your code as follow:

```python

tf.keras.backend.clear_session()
tf.random.set_seed(42)

enc_vocab_size = len(train_vocab) + 1
dec_vocab_size = len(target_vocab) + 1
embed_size = 10


import tensorflow_addons as tfa

encoder_inputs = keras.layers.Input(shape=[None], dtype=np.int32)
decoder_inputs = keras.layers.Input(shape=[None], dtype=np.int32)
sequence_lengths = keras.layers.Input(shape=[], dtype=np.int32)


encoder_embeddings = keras.layers.Embedding(enc_vocab_size, embed_size)(encoder_inputs)
encoder = keras.layers.LSTM(512, return_state = True)
encoder_outputs, state_h, state_c = encoder(encoder_embeddings)
encoder_state = [state_h, state_c]


sampler = tfa.seq2seq.sampler.TrainingSampler()


decoder_embeddings = keras.layers.Embedding(dec_vocab_size, embed_size)(decoder_inputs)
decoder_cell = keras.layers.LSTMCell(512)
output_layer = keras.layers.Dense(dec_vocab_size)



beam_width = 10
start_tokens = tf.zeros([32], tf.dtypes.int32) # assuming 'sos' is zero
end_token = tf.constant(1, tf.dtypes.int32) # assuimg 'eos' is one
decoder = tfa.seq2seq.beam_search_decoder.BeamSearchDecoder(cell = decoder_cell, beam_width = beam_width, output_layer = output_layer)
decoder_initial_state = tfa.seq2seq.beam_search_decoder.tile_batch(encoder_state, multiplier = beam_width)
outputs, _, _ = decoder(decoder_embeddings, start_tokens = start_tokens, end_token = end_token, initial_state = decoder_initial_state)
Y_proba = tf.nn.softmax(outputs.rnn_output)


model = keras.models.Model(inputs = [encoder_inputs, decoder_inputs], outputs = [Y_proba])
model.compile(loss="sparse_categorical_crossentropy", optimizer = 'adam', metrics = ['accuracy'])
```

And got the following error:

```python

ValueError: Exception encountered when calling layer "lstm_cell_1" (type LSTMCell).
    
    Dimensions must be equal, but are 80 and 320 for '{{node beam_search_decoder/decoder/while/BeamSearchDecoderStep/lstm_cell_1/mul}} = Mul[T=DT_FLOAT](beam_search_decoder/decoder/while/BeamSearchDecoderStep/lstm_cell_1/Sigmoid_1, beam_search_decoder/decoder/while/BeamSearchDecoderStep/Reshape_2)' with input shapes: [320,80,2048], [320,512].
    
    Call arguments received:
      • inputs=tf.Tensor(shape=(320, None, 10), dtype=float32)
      • states=ListWrapper(['tf.Tensor(shape=(320, 512), dtype=float32)', 'tf.Tensor(shape=(320, 512), dtype=float32)'])
      • training=None


Call arguments received:
  • embedding=tf.Tensor(shape=(None, None, 10), dtype=float32)
  • start_tokens=tf.Tensor(shape=(32,), dtype=int32)
  • end_token=tf.Tensor(shape=(), dtype=int32)
  • initial_state=['tf.Tensor(shape=(None, 512), dtype=float32)', 'tf.Tensor(shape=(None, 512), dtype=float32)']
  • training=None
  • kwargs=<class 'inspect._empty'>
```

Thanks!

**Comment by kasrahabib** [comment]
Hi @ethenkaufmann,

Here is a minimalistic implementation; without attention, of what you want.
I copied the same setting of encoder decoder from this chapter 16's notebook.
Beamserach is used during inference once we are done training.
I want you to pay close attention to ```beam_serach_inferance_model()```, see how the trained model is replaced with beamsearch.
If the implementation solves your problem, do close this issue!

**Encoder-Decoder Setup**

```python
import tensorflow_addons as tfa

keras.backend.clear_session()
embed_size = 32
lstm_units = 128

encoder_input = keras.layers.Input(shape = [None], dtype = tf.int32)
encoder_embd_layer = keras.layers.Embedding(input_dim = len(INPUT_CHARS) + 1, output_dim = embed_size)
encoder_embd = encoder_embd_layer(encoder_input)
encoder = keras.layers.LSTM(lstm_units, return_sequences = True, return_state = True)
encoder_output, state_h, state_c = encoder(encoder_embd)
encoder_state_HC = [state_h, state_c]

sampler = tfa.seq2seq.TrainingSampler()

decoder_input = keras.layers.Input(shape = [None], dtype = tf.int32)
decoder_embd_layer = keras.layers.Embedding(input_dim = len(OUTPUT_CHARS) + 2, output_dim = embed_size)
decoder_embd = decoder_embd_layer(decoder_input)


LSTMCell = keras.layers.LSTMCell(lstm_units)
output_layer = keras.layers.Dense(len(OUTPUT_CHARS) + 2)

decoder = tfa.seq2seq.BasicDecoder(LSTMCell, sampler, output_layer = output_layer)
final_output, final_state, final_sequence_length = decoder(decoder_embd, initial_state = encoder_state_HC)
y_probas = keras.layers.Activation('softmax')(final_output.rnn_output)

model = keras.Model(inputs = [encoder_input, decoder_input], outputs = [y_probas])
optimizer = keras.optimizers.Nadam()
loss = keras.losses.sparse_categorical_crossentropy
model.compile(loss = loss, optimizer = optimizer, metrics = ['accuracy'])
history = model.fit([X_train, X_train_decoder], Y_train, validation_data = ([X_valid, X_valid_decoder], Y_valid), epochs = 1)
```

**Implementing Beamsearch**
```python
def beam_serach_inferance_model(beam_width):
  batch_size = tf.shape(encoder_input)[:1]
  max_output_length = Y_train.shape[1]
  start_tokens = tf.fill(dims = batch_size, value = sos_id)
  decoder_initial_state = tfa.seq2seq.tile_batch(encoder_state_HC, multiplier = beam_width)
  beam_search_inference = tfa.seq2seq.BeamSearchDecoder(cell = LSTMCell, beam_width = beam_width, output_layer = output_layer, maximum_iterations = max_output_length)
  outputs, _, _ = beam_search_inference(decoder_embd_layer.variables, start_tokens = start_tokens, end_token = 0, initial_state = decoder_initial_state)
  final_outputs = tf.transpose(outputs.predicted_ids, perm = (0,2,1))
  beam_scores = tf.transpose(outputs.beam_search_decoder_output.scores, perm = (0,2,1))
  return keras.Model(inputs = [encoder_input], outputs = [final_outputs, beam_scores])

beam_serach_inferance_model = beam_serach_inferance_model(3)
```

**Utility function** 
*I copied this function from TFA's API tutorial and adapted it!!!*

```python
def beam_translate(sentence):
  X = prepare_date_strs_padded(sentence)
  result, beam_scores = beam_serach_inferance_model.predict(X)
  for beam, score in zip(result, beam_scores):
    output = ids_to_date_strs(beam)
    beam_score = [a.sum() for a in score]
    print('Input: %s' % sentence)
    print('-----' * 12)
    for i in range(len(output)):
      print('{} Predicted translation: {}  {}'.format(i + 1, output[i], beam_score[i]))
    print('\n')
```

**Output**

```python
beam_translate(["July 14, 1789", "September 01, 2020"])

Input: ['July 14, 1789', 'September 01, 2020']
------------------------------------------------------------
1 Predicted translation: 2288-01-11  -83.7786865234375
2 Predicted translation: 2288-01-10  -83.90345764160156
3 Predicted translation: 2288-01-21  -84.30797576904297


Input: ['July 14, 1789', 'September 01, 2020']
------------------------------------------------------------
1 Predicted translation: 2221-02-26  -79.02340698242188
2 Predicted translation: 2222-02-26  -79.29275512695312
3 Predicted translation: 2221-02-21  -80.06587982177734
```

I hope this helps!

Hi @ageron,

I know the notebook is too long, but I think it would be nice to have a few lines on beamsearch.
Moreover, I really liked your implementation of ```Multi-Headed Attention```, all vectorized! 👍☺️


Cheers,
Kasra

**Comment by ethenkaufmann** [comment]
Thanks!
This is helpful, I will close the issue!

**Comment by ageron** [comment]
Hi @ethenkaufmann ,
I added a section on beam search in the upcoming 3rd edition. You can access the early release at https://homl.info/er3, and the notebook at https://colab.research.google.com/github/ageron/handson-ml3/blob/main/16_nlp_with_rnns_and_attention.ipynb
There's a basic implementation of beam search in that notebook.
Hope this helps
